orc8r/cloud/go/services/metricsd/prometheus/prometheus-cache/cache/cache.go

```diff
@@ -147,18 +148,51 @@ func (c *MetricCache) clearMetrics() {
 	c.metricFamiliesByName = make(map[string]*familyAndMetrics)
 }
 
-func (c *MetricCache) exposeMetrics(metricFamiliesByName map[string]*familyAndMetrics) string {
-	respStr := strings.Builder{}
+func (c *MetricCache) exposeMetrics(metricFamiliesByName map[string]*familyAndMetrics, workers int) string {
+	fams := make(chan *familyAndMetrics, workers)
+	results := make(chan string, workers)
+	respStr := make(chan string, 1)
+
+	waitGroup := &sync.WaitGroup{}
+
+	for i := 0; i < workers; i++ {
+		waitGroup.Add(1)
+		go processFamilyWorker(fams, results, waitGroup)
+	}
+
+	go processFamilyStringsWorker(results, respStr)
+
 	for _, fam := range metricFamiliesByName {
+		fams <- fam
+	}
+
+	close(fams)
+	waitGroup.Wait()
+	close(results)
+
+	return <-respStr
+}
+
+func processFamilyWorker(fams <-chan *familyAndMetrics, results chan<- string, waitGroup *sync.WaitGroup) {
+	defer waitGroup.Done()
+	for fam := range fams {
```

Comment:
if you close the `fams` channel before workers are done, you'll pull a `nil` off the channel here. The body of the loop here should be nullsafe

Reply:
As far as I understand, the close simply says no more values will be added on it. This allows the for fam := range fams {} to terminate. It doesn't cause nil to be pulled off. These are the two examples I used:

https://gobyexample.com/worker-pools

https://gobyexample.com/closing-channels

Reply:
Aditya is right